src/mono/mono/mini/jit-icalls.c

```diff
@@ -1449,8 +1449,8 @@ mono_gsharedvt_constrained_call (gpointer mp, MonoMethod *cmethod, MonoClass *kl
 		m = info->method;
 		break;
 	default:
-		/* Object.GetType () is an intrinsic under netcore */
-		if (!mono_class_is_ginst (cmethod->klass) && !cmethod->is_inflated && !strcmp (cmethod->name, "GetType")) {
```

Comment:
Yes, here is the reasoning behind this change. Before the optimization of constrained calls from gshared methods in https://github.com/dotnet/runtime/pull/79339, the type was retrieved as `vt = mono_class_vtable_checked (klass, error);`, in the same way as I did here. After that, in https://github.com/dotnet/runtime/pull/65126 arguments of static virtual methods are handled in the same way as for `Object.GetType ()`:
https://github.com/dotnet/runtime/blob/5d00fc5a1423fa863df853a75f840c8267bead19/src/mono/mono/mini/method-to-ir.c#L3857-L3858

Additionally, there is a note that the mp is null in case of static virtual methods, which is why the `constrained_gsharedvt_call_setup` fails:
https://github.com/dotnet/runtime/blob/5d00fc5a1423fa863df853a75f840c8267bead19/src/mono/mono/mini/jit-icalls.c#L1423-L1433

My assumption was that it shouldn't reach the `constrained_gsharedvt_call_setup` in this case.

Reply:
After offline discussion, the above ^ was wild assumption :) The static methods should be handled in constrained_gsharedvt_call_setup as @BrzVlad suggested.